codec-xml/src/main/java/io/netty/handler/codec/xml/XmlDecoder.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.xml;
+
+import com.fasterxml.aalto.AsyncInputFeeder;
+import com.fasterxml.aalto.AsyncXMLInputFactory;
+import com.fasterxml.aalto.AsyncXMLStreamReader;
+import com.fasterxml.aalto.stax.InputFactoryImpl;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import javax.xml.stream.XMLStreamConstants;
+import java.util.List;
+
+/**
+ * Async XML decoder based on <a href="https://github.com/FasterXML/aalto-xml">Aalto XML parser</a>.
+ *
+ * Parses the incoming data into one of XML messages defined in this package.
+ */
+
+public class XmlDecoder extends ByteToMessageDecoder {
+
+    private static final AsyncXMLInputFactory XML_INPUT_FACTORY = new InputFactoryImpl();
+    private static final XmlDocumentEnd XML_DOCUMENT_END = new XmlDocumentEnd();
+
+    private final AsyncXMLStreamReader streamReader = XML_INPUT_FACTORY.createAsyncXMLStreamReader();
+    private final AsyncInputFeeder streamFeeder = streamReader.getInputFeeder();
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        byte[] buffer = new byte[in.readableBytes()];
+        in.readBytes(buffer);
+        streamFeeder.feedInput(buffer, 0, buffer.length);
```

Comment:
What would happen when an exception occurs?  When an exception is raised, please rethrow the exception and make sure all incoming traffic is discarded.  For an example, take a look at `FastLzFrameDecoder.java` and its `CORRUPTED` state.